libsignal-service-actix/src/push_service.rs

```diff
@@ -210,24 +293,76 @@ impl PushService for AwcPushService {
         }
     }
 
-    async fn get_from_cdn(
+    async fn get_protobuf<T>(
         &mut self,
-        cdn_id: u32,
+        endpoint: Endpoint,
         path: &str,
-    ) -> Result<Self::ByteStream, ServiceError> {
-        use futures::stream::TryStreamExt;
+        credentials: Option<HttpCredentials>,
+    ) -> Result<T, ServiceError>
+    where
+        T: Default + ProtobufMessage,
+    {
+        let mut response = self
+            .request(Method::GET, endpoint, path, credentials)?
+            .send()
+            .await
+            .map_err(|e| ServiceError::SendError {
+                reason: e.to_string(),
+            })?;
 
-        let url = Url::parse(&self.cfg.cdn_urls[&cdn_id])
-            .expect("valid cdn base url")
-            .join(path)
-            .expect("valid CDN path");
+        let text =
+            response
+                .body()
+                .await
+                .map_err(|e| ServiceError::ResponseError {
+                    reason: e.to_string(),
+                })?;
+        Ok(T::decode(text)?)
+    }
+
+    async fn put_protobuf<D, S>(
+        &mut self,
+        endpoint: Endpoint,
+        path: &str,
+        value: S,
+    ) -> Result<D, ServiceError>
+    where
+        D: Default + ProtobufMessage,
+        S: Sized + ProtobufMessage,
+    {
+        let mut buf = vec![];
+        value.encode(&mut buf)?;
```

Comment:
`with_capacity` and `.expect()` until we have https://github.com/danburkert/prost/pull/378